sw/airborne/modules/gps/gps_datalink.c

```diff
@@ -66,78 +67,42 @@ void gps_datalink_init(void)
-  AbiSendMsgIMU_MAG_INT32(MAG_DATALINK_SENDER_ID, now_ts, &mag);
-}
+  struct EnuCoor_i enu_pos_i, enu_speed_i;
 
-// Parse the REMOTE_GPS_SMALL datalink packet
-static void parse_gps_datalink_small(int16_t heading, uint32_t pos_xyz, uint32_t speed_xyz, uint32_t tow)
-{
-  struct EnuCoor_i enu_pos, enu_speed;
-
-  // Position in ENU coordinates
-  enu_pos.x = (int32_t)((pos_xyz >> 21) & 0x7FF); // bits 31-21 x position in cm
-  if (enu_pos.x & 0x400) {
-    enu_pos.x |= 0xFFFFF800;  // sign extend for twos complements
-  }
-  enu_pos.y = (int32_t)((pos_xyz >> 10) & 0x7FF); // bits 20-10 y position in cm
-  if (enu_pos.y & 0x400) {
-    enu_pos.y |= 0xFFFFF800;  // sign extend for twos complements
-  }
-  enu_pos.z = (int32_t)(pos_xyz & 0x3FF); // bits 9-0 z position in cm
+  enu_pos_i.x = (int32_t)CM_OF_M(enu_pos->x);
+  enu_pos_i.y = (int32_t)CM_OF_M(enu_pos->y);
+  enu_pos_i.z = (int32_t)CM_OF_M(enu_pos->z);
 
   // Convert the ENU coordinates to ECEF
-  ecef_of_enu_point_i(&gps_datalink.ecef_pos, &ltp_def, &enu_pos);
+  ecef_of_enu_point_i(&gps_datalink.ecef_pos, &ltp_def, &enu_pos_i);
   SetBit(gps_datalink.valid_fields, GPS_VALID_POS_ECEF_BIT);
 
   lla_of_ecef_i(&gps_datalink.lla_pos, &gps_datalink.ecef_pos);
   SetBit(gps_datalink.valid_fields, GPS_VALID_POS_LLA_BIT);
 
-  enu_speed.x = (int32_t)((speed_xyz >> 21) & 0x7FF); // bits 31-21 speed x in cm/s
-  if (enu_speed.x & 0x400) {
-    enu_speed.x |= 0xFFFFF800;  // sign extend for twos complements
-  }
-  enu_speed.y = (int32_t)((speed_xyz >> 10) & 0x7FF); // bits 20-10 speed y in cm/s
-  if (enu_speed.y & 0x400) {
-    enu_speed.y |= 0xFFFFF800;  // sign extend for twos complements
-  }
-  enu_speed.z = (int32_t)((speed_xyz) & 0x3FF); // bits 9-0 speed z in cm/s
-  if (enu_speed.z & 0x200) {
-    enu_speed.z |= 0xFFFFFC00;  // sign extend for twos complements
-  }
+  enu_speed_i.x = (int32_t)CM_OF_M(enu_speed->x);
+  enu_speed_i.y = (int32_t)CM_OF_M(enu_speed->y);
+  enu_speed_i.z = (int32_t)CM_OF_M(enu_speed->z);
 
-  VECT3_NED_OF_ENU(gps_datalink.ned_vel, enu_speed);
+  VECT3_NED_OF_ENU(gps_datalink.ned_vel, enu_speed_i);
   SetBit(gps_datalink.valid_fields, GPS_VALID_VEL_NED_BIT);
 
-  ecef_of_enu_vect_i(&gps_datalink.ecef_vel , &ltp_def , &enu_speed);
+  ecef_of_enu_vect_i(&gps_datalink.ecef_vel , &ltp_def , &enu_speed_i);
   SetBit(gps_datalink.valid_fields, GPS_VALID_VEL_ECEF_BIT);
 
-  gps_datalink.gspeed = (int16_t)FLOAT_VECT2_NORM(enu_speed);
-  gps_datalink.speed_3d = (int16_t)FLOAT_VECT3_NORM(enu_speed);
+  gps_datalink.gspeed = (int16_t)FLOAT_VECT2_NORM(enu_speed_i);
+  gps_datalink.speed_3d = (int16_t)FLOAT_VECT3_NORM(enu_speed_i);
 
-  gps_datalink.hmsl = ltp_def.hmsl + enu_pos.z * 10;
+  gps_datalink.hmsl = ltp_def.hmsl + enu_pos_i.z * 10;
   SetBit(gps_datalink.valid_fields, GPS_VALID_HMSL_BIT);
 
-  gps_datalink.course = ((int32_t)heading) * 1e3;
+  gps_datalink.course = (int32_t)(RadOfDeg(course)*1e7);
```

Comment:
To avoid multiple transformation, the course in this function should be already in radians
```suggestion
  gps_datalink.course = (int32_t)(course*1e7);
```